networkx/algorithms/summarization.py

```diff
@@ -210,3 +210,456 @@ def dedensify(G, threshold, prefix=None, copy=True):
+                    edge_types.append(edge_type_attributes)
+
+                has_edges = output.has_edge(*summary_edge)
+                if output.is_multigraph():
+                    existing_edges = output.get_edge_data(*summary_edge)
+                    for edge_type in edge_types:
+                        if not has_edges or (
+                            not output.is_directed()
+                            and not any(
+                                edge_type == data
+                                for edge, data in existing_edges.items()
+                            )
+                        ):
+                            output.add_edge(*summary_edge, **edge_type)
+                else:
+                    superedge_attributes = {superedge_attribute: edge_types}
+                    if not output.is_directed():
+                        if (
+                            has_edges
+                            and output.get_edge_data(*summary_edge)
+                            == superedge_attributes
+                        ):
+                            continue
+
+                    output.add_edge(*summary_edge, **superedge_attributes)
+
+    nx.set_node_attributes(output, supernode_lookup, supernode_attribute)
+    return output
+
+
+def snap_build_data_structures(G, node_attributes, edge_attributes):
+    """Returns data structures needed to build the SNAP summary graph
+
+    Creates the data structures and splits the node groups  until all are groups
+    are homogenous in terms of node attribute values and edge types to
+    build the output graph for the SNAP summary graph.
+
+    Parameters
+    ----------
+    G: graph
+        Networkx Graph to be summarized
+    node_attributes: iterable
+        The node attributes that can be found in nodes in the graph, and are to
+        be recognized in the summarization
+    edge_attributes: iterable
+        The edge attributes that can be found in edges in the graph, and
+        are to be recognized in the summarization
+
+    Returns
+    -------
+    tuple of data structure representations: (list, list, dict, dict)
+
+    Notes
+    -----
+    Finds the AR-compatible grouping of graph `G` and returns data structures
+    to efficiently process those groupings.
+
+    An Attribute-edge compatible (AR-compatible) grouping is where
+    all nodes in each group have the same exact node attribute values and
+    have the same exact edge types to nodes in the same exact
+    groups.
+    """
+    edge_types = {
+        edge: tuple(attrs.get(attr) for attr in edge_attributes)
+        for edge, attrs in G.edges.items()
+    }
+    if not G.is_directed():
+        if G.is_multigraph():
+            edge_types.update(
+                ((e[1], e[0], e[2]), edge_types[e]) for e in list(edge_types)
+            )
+        else:
+            edge_types.update(
+                ((v, u), edge_types[(u, v)]) for (u, v) in list(edge_types)
+            )
```

Comment:
```suggestion
        if G.is_multigraph():
            # list is needed to avoid mutating while iterating
            edges = [((v, u, k), etype) for (u, v, k), etype in edge_types.items()]
        else:
            # list is needed to avoid mutating while iterating
            edges = [((v, u), etype) for (u, v), etype in edge_types.items()]
        edge_types.update(edges)
```